rio/src/RIO/File.hs

```diff
@@ -0,0 +1,345 @@
+        fsyncDescriptor "File" (FD.fdFD fileFd) `finally` hClose h
+        fsyncDescriptor "Directory" dirFd)
+    closeDirectory
+  where
+    fsyncDescriptor name fd =
+      void $
+      throwErrnoIfMinus1 ("closeFileDurable - fsync(" <> name <> ")") $
+      c_safe_fsync fd
+    closeDirectory =
+      void $
+      throwErrnoIfMinus1Retry "closeFileDurable - close(Directory)" $
+      c_close dirFd
+
+toTmpFilePath :: FilePath -> FilePath
+toTmpFilePath filePath =
+    dirPath </> tmpFilename
+  where
+    dirPath = takeDirectory filePath
+    filename = takeFileName filePath
+    tmpFilename = "." <> filename <> ".tmp"
+
+
+
+handleToFd :: Handle -> IO FD.FD
+#if MIN_VERSION_base(4,9,0)
+handleToFd h =
+  case h of
+    HandleFD.FileHandle _ mv -> do
+      HandleFD.Handle__{HandleFD.haDevice = dev} <- readMVar mv
+      case cast dev of
+        Just fd -> return fd
+        Nothing -> error "not a file handle"
+    HandleFD.DuplexHandle {} -> error "not a file handle"
+# else
+handleToFd = HandleFD.handleToFd
+#endif
+
+-- | This sub-routine does the following tasks:
+--
+-- * It calls fsync and then closes the given handler (mapping to a temporal/backup filepath)
+-- * It renames the file to the original path (using renameat)
+-- * It calls fsync and then closes the containing directory of the file
+--
+-- These steps guarante that the file are durable, and that the backup mechanism
+-- for catastrophic failure is discarded after no error is thrown.
+--
+-- @since 0.1.6
+closeFileDurableAtomic ::
+     MonadIO m => FilePath -> Fd -> Handle -> m ()
+closeFileDurableAtomic filePath (Fd dirFd) fileHandler =
+  liftIO $
+  finally
+    (withFilePath tmpFilePath $ \tmpFp ->
+       withFilePath filePath $ \fp -> do
+         fileFd <- handleToFd fileHandler
+         fsyncDescriptor "File" (FD.fdFD fileFd)
+           `finally` hClose fileHandler
+         renameFile tmpFp fp
+         fsyncDescriptor "Directory" dirFd)
+    closeDirectory
+  where
+    tmpFilePath = toTmpFilePath filePath
+    fsyncDescriptor name fd =
+      void $
+      throwErrnoIfMinus1 ("closeFileDurableAtomic - fsync(" <> name <> ")") $
+      c_safe_fsync fd
+    renameFile tmpFp origFp =
+      void $
+      throwErrnoIfMinus1Retry "closeFileDurableAtomic - renameFile" $
+      c_safe_renameat dirFd tmpFp dirFd origFp
+    closeDirectory =
+      void $
+      throwErrnoIfMinus1Retry "closeFileDurableAtomic - close(Directory)" $
+      c_close dirFd
+
+#endif
+
+-- | After a file is closed, it opens it again and executes fsync internally on
+-- both the file and the directory that contains it.
+--
+-- [The effectiveness of calling this function is
+-- debatable](https://stackoverflow.com/questions/37288453/calling-fsync2-after-close2/50158433#50158433),
+-- as this relies on internal implementation details at the Kernel level that
+-- might change. We argue that, despite this fact, calling this function may
+-- bring benefits in terms of durability
```

Comment:
Is this intended to be used together with the other durable functions in this module, or with non-durable file writing functions?

Reply:
Is intended to be used with non-durable file writing functions. The idea is if you have a call in your module that doesn't enforce durability (fsync), calling this function after that would at least attempt to fsync after it was closed the first time. There is no 100% guarantee this will work, but in some systems, it may...

Reply:
No objection, but please clarify that this is intended to work around non-durability in other APIs, as opposed to being necessary for the API functions provided here.